Display files hierarchically in Web UI content tab

Piccirello opened the pull request:
Rename functionality is going to take some more time so will be coming in a later PR.

- [x] Minor API bump to document `/api/v2/torrents/files` path separator change

Closes #974, closes #4476, closes #10162, closes #10484, closes #10729

Before:
<img width="1116" alt="screen shot 2019-01-13 at 4 22 23 am" src="https://user-images.githubusercontent.com/8296030/51085296-c6cd9c00-16eb-11e9-96ca-7c36d50853b7.png">

After:
<img width="1072" alt="screen shot 2019-01-13 at 4 20 48 am" src="https://user-images.githubusercontent.com/8296030/51085298-c9c88c80-16eb-11e9-8249-4233438385d5.png">

Piccirello commented:
Added @xnoreq's patch for #10162

Piccirello commented:
Rebased to fix conflicts

Piccirello commented on .gitignore:
```diff
@@ -35,3 +35,7 @@ config.status
 src/icons/skin/build-icons/node_modules/
 src/icons/skin/build-icons/icons/*.png
+
+# Web UI tools
```
Added this for installing js-beautify locally

Chocobo1 commented on src/webui/www/private/scripts/client.js:
```diff
@@ -778,6 +789,14 @@ window.addEvent('load', function() {
+        $('torrentFilesFilterToolbar').removeClass("invisible");
+    };
+
+    var hideFilesFilter = function(bool) {
```
either merge the 2 functions or remove the `bool` parameter.

Chocobo1 commented on src/webui/www/private/scripts/fs.js:
```diff
@@ -0,0 +1,46 @@
+'use strict';
```
license missing.

Chocobo1 commented on src/webui/www/private/scripts/fs.js:
```diff
@@ -0,0 +1,46 @@
+    }
+}).send();
+
+// This file is the JavaScript implementation of base/utils/fs.cpp
```
this should be at top of this file.

Chocobo1 commented on src/webui/www/private/scripts/fs.js:
```diff
@@ -0,0 +1,46 @@
+function fileExtension(filename) {
+    var name = filename.replace(QB_EXT, '');
+    var pointIndex = name.lastIndexOf('.');
+    if (pointIndex >= 0)
```
I would prefer to write `(pointIndex === -1)` to have the same style with the next 2 functions.

Chocobo1 commented on src/webui/www/private/scripts/fs.js:
```diff
@@ -0,0 +1,46 @@
+ * Returns the file extension part of a file name.
+ */
+function fileExtension(filename) {
+    var name = filename.replace(QB_EXT, '');
```
i think it is better to check with endsWith() then trim it if found.

Chocobo1 commented on src/webui/www/private/scripts/dynamicTable.js:
```diff
@@ -1586,6 +1586,35 @@ var TorrentTrackersTable = new Class({
     Extends: DynamicTable,
 
+    rootNode: null,
+    rowToNodeMap: new Hash(),
```
unused variable.

Chocobo1 replied:
why not `rowToNodeMap: {},`?

Chocobo1 commented on src/webui/www/private/scripts/dynamicTable.js:
```diff
@@ -1649,11 +1733,97 @@ var TorrentFilesTable = new Class({
+            nodes.sort(function(row1, row2) {
+                var column = this.columns[this.sortedColumn];
+                var res = column.compareRows(row1, row2);
+                if (this.reverseSort == '0')
```
`===` possible?

Chocobo1 replied:
or `return (this.reverseSort === '0') ? res : -res;`

Chocobo1 commented on src/webui/www/private/scripts/dynamicTable.js:
```diff
@@ -1649,11 +1733,97 @@ var TorrentFilesTable = new Class({
+            }.bind(this));
+
+            nodes.each(function(node) {
+                if (node.children.length)
```
`node.children.length > 0`

Chocobo1 commented on src/webui/www/private/scripts/dynamicTable.js:
```diff
@@ -1649,11 +1733,97 @@ var TorrentFilesTable = new Class({
+        }.bind(this);
+
+        var isFiltered = function(name) {
+            for (var i = 0; i < filterTerms.length; ++i) {
```
how about?
```js
function(name) {
    return filterTerms.every(function(term) {
        return (name.indexOf(term) !== -1);
    });
};
```
Besides that, is it possible to just use a bool to store the filtered state?

Piccirello replied:
I've pushed new changes that "cache" the filtered set.

Chocobo1 commented on src/webui/www/private/scripts/dynamicTable.js:
```diff
@@ -1649,11 +1733,97 @@ var TorrentFilesTable = new Class({
+
+        var filterNodes = function(node) {
+            if (node.isFolder) {
+                var childAdded = false;
```
Just a suggestion:
```js
var childAdded = node.children.reduce(function(acc, child) {
    return (acc | filterNodes(child));
}, false);
```
ah, the joy of functional programming...

Piccirello commented on src/webui/api/appcontroller.h:
```diff
@@ -47,4 +47,5 @@ private slots:
     void preferencesAction();
     void setPreferencesAction();
     void defaultSavePathAction();
+    void pathSeparatorAction();
```
That would be ideal. The api currently converts all file names to native paths before sending: [source](https://github.com/qbittorrent/qBittorrent/blob/84e683cc999776c7b1f521db0bb9381d253b1f2c/src/webui/api/torrentscontroller.cpp#L463).

glassez replied:
>That would be ideal. 

Then why don't you go that way instead of sending server internals? Just change appropriate code.

Piccirello replied:
Are you confident that won't break anything? I'm not.

glassez replied:
What can it break?
Just try to analyze it. Currently it has undefined behavior from the client side perspective (the client does not know exactly which separator is used). So if we normalize this behavior as suggested above, the old client code will not even feel it, but the new one will only benefit.

Piccirello replied:
I'll make the change.

Do we have a sense of when 4.2.0 will be released? I want this PR to be thoroughly tested before it makes it into a release.

glassez replied:
>I want this PR to be thoroughly tested before it makes it into a release.

:+1:
>Do we have a sense of when 4.2.0 will be released?

Personally, I would wait for it to fully support the libtorrent-1.2 (in native mode, i.e. with deprecated features disabled). Anyway, I think you have enough time for testing.

Arathen commented:
Is there an update on this PR by any chance? I would love to see it merged.

Cheers.

Piccirello commented:
I’ve been working on a large overhaul of this PR over the last week. It addresses all comments and cleans up the code quite a bit. There’s still a few todos but hoping to have it all pushed within the next few days.

WolfganP commented:
I've recently started to see the Download Priority boxes being grayed out when doing selective downloads and putting the torrent in pause (ie getting just the 1st 2 episodes of a series' season). 
In order to recover the functionality of that field and be able to change the priority for individual files, I need to restart the program. I'll make sure I test it again once this PR is merged.

Piccirello commented:
@WolfganP have you been experiencing this issue while running off of master, or running off of this branch?

WolfganP commented:
> 
> 
> @WolfganP have you been experiencing this issue while running off of master, or running off of this branch?

@Piccirello Always running latest master

Piccirello commented:
My latest changes have been pushed. This should probably be merged after #10527 due to a large number of impending merge conflicts.

glassez commented:
@Piccirello, please extract unrelated changes into separate PR (I mean changes in qBittorrent GUI files).

Piccirello commented:
GUI changes moved to #10839

Chocobo1 commented on src/webui/www/webui.qrc:
```diff
@@ -37,6 +38,7 @@
         <file>private/scripts/mocha-init.js</file>
         <file>private/scripts/progressbar.js</file>
         <file>private/scripts/prop-files.js</file>
+        <file>private/scripts/file-tree.js</file>
```
wrong place, probably a few lines up.

Piccirello commented:
I want to manually rebase this before we merge

jonfortescue commented:
@Piccirello looks like this can be rebased now that #10527 has merged through.

Piccirello commented:
Rebased onto master

Chocobo1 commented on src/webui/www/private/scripts/dynamicTable.js:
```diff
@@ -1712,7 +1829,7 @@ const TorrentFilesTable = new Class({
             if (progressBar === null) {
                 td.adopt(new ProgressBar(value.toFloat(), {
-                    'id': 'pbf_' + id,
+                    'id': `pbf_${id}`,
```
This is ES2015 template literals right? IE doesn't support it.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#Browser_compatibility

Piccirello replied:
Good catch, will switch them back

Chocobo1 commented on src/webui/www/private/scripts/dynamicTable.js:
```diff
@@ -1728,11 +1845,155 @@ const TorrentFilesTable = new Class({
+                return (acc || (term !== this.prevFilterTerms[index]));
+            }.bind(this), false);
+            const isFilterChanged = ((this.filterTerms.length !== this.prevFilterTerms.length)
+                || ((this.filterTerms.length !== 0) && isFilterTermsChanged));
```
maybe use `> 0` instead of `!== 0`?

Chocobo1 commented on src/webui/www/private/scripts/dynamicTable.js:
```diff
@@ -1687,6 +1748,60 @@ const TorrentFilesTable = new Class({
 
+        this.columns['name'].updateTd = function(td, row) {
+            const id = row.rowId;
+            const fileNameId = `filesTablefileName${id}`;
```
ES2015 template literals? more below.

Piccirello commented:
Comments addressed

Chocobo1 commented on src/webui/www/private/scripts/file-tree.js:
```diff
@@ -0,0 +1,221 @@
+'use strict';
+
+const FilePriority = {
+  "Ignored": 0,
```
we use 4 spaces for indention.

Chocobo1 commented on src/webui/www/private/scripts/file-tree.js:
```diff
@@ -0,0 +1,221 @@
+};
+Object.freeze(TriState);
+
+const FileTree = new Class({
```
where is this `Class` from? seems different to ECMAScript 2015 classes.

Piccirello replied:
This is a mootools class as IE11 doesn't support ES6 classes

Chocobo1 commented on src/webui/www/webui.qrc:
```diff
@@ -30,6 +30,8 @@
         <file>private/scripts/download.js</file>
         <file>private/scripts/dynamicTable.js</file>
+        <file>private/scripts/file-tree.js</file>
+        <file>private/scripts/fs.js</file>
```
I would just use the full name "filesystem.js" leaving no ambiguity.

Piccirello commented:
Comments addressed

Chocobo1 commented:
@Piccirello 
Thank you!

Arathen commented:
@Piccirello - thank you so much for your long hard work on this PR! Its a fantastic addition to the WebUI :)

Piccirello commented:
@Arathen I'm glad this is finally implemented! The original issue is almost 6 years old.

glassez commented:
> I'm glad this is finally implemented! The original issue is almost 6 years old.

Maybe you can turn your attention to an equally old issue now?
#453

foremtehan commented:
How can i get list of torrents files before downloading the torrent itself ? Just see the content (metadata) of the torrent?
